tactics/rewrite.ml

```diff
@@ -1083,6 +1083,22 @@ let subterm all flags (s : 'a pure_strategy) : 'a pure_strategy =
                     in state, res
             else rewrite_args state None
 
+      | Proj (p, r, arg) ->
+          let expanded = Retyping.expand_projection env (fst evars) p arg [] in
+          let unfold_compat_constant sigma t =
+            match EConstr.kind sigma t with
+            | App (id, arr) -> mkProj (p, r, Array.last arr)
+            | _ -> assert false in
+            let state, res = s.strategy { state ; env ; unfresh ;
+                                 term1 = expanded ; ty1 = ty ;
+                                 cstr = (prop,cstr) ; evars = evars } in
+          let res =
+            match res with
+            | Success r -> Success { r with rew_to = unfold_compat_constant (goalevars r.rew_evars) r.rew_to }
```

Comment:
You should try moving the call to expand_projection inside the pattern-matching clause or into the closure.